### PR TITLE
[rllib] Parallelize sample collection and gradient computation in DQN

### PR DESCRIPTION
**Pull request by ericl**
Refactor DQN to use a number of worker processes, each of which has an independent replay buffer. This allows sample collection and gradient computation to proceed in parallel. In each iteration, the driver applies worker gradients and re-broadcasts them to workers.

There is no parallelization by default, since it does not produce speedups on simple problems. This will be addressed as a follow-up.

### CODE REVIEW

File: python/ray/rllib/dqn/dqn.py [hunk 1]
```diff
@@ -124,92 +189,107 @@ def make_obs_ph(name):
...
     self.num_iterations = 0
+    self.num_target_updates = 0
+    self.steps_since_update = 0
+
+  def _get_rollouts(self, steps_requested, cur_timestep):
+    requests_total = 0
+    requests = {}
+
+    def new_request(worker):
+      nonlocal requests_total
```
Comment (robertnishihara) [inline python/ray/rllib/dqn/dqn.py]: this only works in python 3, right?

**Comment by robertnishihara** [comment]
I attempted to rebase it at https://github.com/ray-project/ray/compare/master...robertnishihara:dqn-parallel-rebased. I didn't force push in case I messed something up.

**Comment by ericl** [comment]
This is going to need rebasing once the other DQN refactoring pr goes in,
I'll take care of that since there will be some conflicts.

**Comment by ericl** [comment]
@pcmoritz how about we try to get this merged, it's a decent sized refactoring and causes conflicts often.

**Pull request merged**